### PR TITLE
New abscal phase solver

### PR DESCRIPTION
**Pull request by tyler-a-cox**
This PR folds in @zacharymartinot new abscal phase gradient solver. The new solver uses FFTs in position space to solve for the per-frequency phase gradient, and seems much less sensitive to large errors in phase than the existing methods.

### CODE REVIEW

File: hera_cal/abscal.py
Comment (jsdillon) [inline hera_cal/abscal.py:1563]: For ease of reading and logical flow, I'd put this line with the previous block

File: hera_cal/abscal.py
Comment (jsdillon) [inline hera_cal/abscal.py:1569]: Question for @zacharymartinot: is there anything else that's not too memory intensive that we want to pass through via this meta dict?

File: hera_cal/abscal.py
Comment (jsdillon) [inline hera_cal/abscal.py:1345]: Might be good to add comments (or more to the docstring) explaining the analytics here and why these sines and cosines are the right thing to do.

File: hera_cal/abscal.py
Comment (jsdillon) [inline hera_cal/abscal.py:1389]: Question for @zacharymartinot: Is there any reason to step less than -1 * the hessian (i.e. a damped iterator?)

File: hera_cal/abscal.py
Comment (jsdillon) [inline hera_cal/abscal.py:1516]: This should clarify that it only performs absolute calibration of phase slopes, not amplitudes.

File: hera_cal/tests/test_abscal.py
Comment (jsdillon) [inline hera_cal/tests/test_abscal.py:727]: In theory, you shouldn't have to use the transformed vectors when "uncalibrating" here.

Also, it might be worth testing this on an array that has extra DoF... perhaps two widely separated hexagons, with a phase difference between then in addition to EW and NS phase gradients.

**Review by jsdillon — CHANGES_REQUESTED** [review]
A few minor clarifications, but this is looking really good @tyler-a-cox. Thanks for cleaning this up, adding more documentation, and writing unit tests!

I'd like to get @zacharymartinot's go-ahead before we merge in.

File: hera_cal/tests/test_abscal.py
Comment (tyler-a-cox) [reply hera_cal/tests/test_abscal.py:727]: Good point, I'll add that test.

File: hera_cal/abscal.py
Comment (tyler-a-cox) [reply hera_cal/abscal.py:1563]: Yeah, that makes sense. I'll move that line up

File: hera_cal/abscal.py
Comment (tyler-a-cox) [reply hera_cal/abscal.py:1516]: I'll make that change

File: hera_cal/abscal.py
Comment (tyler-a-cox) [reply hera_cal/abscal.py:1389]: I did try this in another implementation of this function. I could just make the change and add it as a keyword argument.

File: hera_cal/abscal.py
Comment (tyler-a-cox) [reply hera_cal/abscal.py:1345]: Will do

File: hera_cal/abscal.py
Comment (zacharymartinot) [reply hera_cal/abscal.py:1389]: I thought about this briefly, but not very carefully. Its possible that including a damping factor could be useful when we're at the hairy edge of the model being "good enough". But I'm pretty sure a constant damping factor isn't going to do anything than add iterations in general, it would have to be something computed per step. and figuring that out would be a small project, if there's anything there at all. And if you're going down that road you might as well consider alternatives to newton's method entirely.

So until someone actually investigates, keep it simple. No damping.

Comment (jsdillon) [reply hera_cal/abscal.py:1389]: Yeah, sounds like an unnecessary rabbit hole... just something I thought I'd mention.

Comment (zacharymartinot) [reply hera_cal/abscal.py:1389]: I thought about this briefly, but not very carefully. Its possible that including a damping factor could be useful when we're at the hairy edge of the model being "good enough". But I'm pretty sure a constant damping factor isn't going to do anything than add iterations in general, it would have to be something computed per step. and figuring that out would be a small project, if there's anything there at all. And if you're going down that road you might as well consider alternatives to newton's method entirely.

So until someone actually investigates, keep it simple. No damping.

File: hera_cal/abscal.py
Comment (zacharymartinot) [reply hera_cal/abscal.py:1569]: I added collection of the number of newton's method iterations. We should consider flagging on hitting maxiter.

File: hera_cal/abscal.py
Comment (zacharymartinot) [reply hera_cal/abscal.py]: This bit is gone now.

File: hera_cal/abscal.py
Comment (zacharymartinot) [reply hera_cal/abscal.py]: avoiding evaluating complex exponentials/creating complex arrays with a little trig

File: hera_cal/tests/test_abscal.py
Comment (zacharymartinot) [reply hera_cal/tests/test_abscal.py:727]: I think that using the transformed vectors to uncalibrate is a fine unit test. And if we just add `split_core=True` then the test array will have 4 degeneracies.

Comment (zacharymartinot) [reply hera_cal/tests/test_abscal.py:728]: I changed the test array to have 4 degeneracies.

File: hera_cal/abscal.py
Comment (zacharymartinot) [reply hera_cal/abscal.py]: Done.

**Review by jsdillon — APPROVED** [review]
Ok, I'm good to go!

**Review by zacharymartinot — APPROVED** [review]
Looks good to me. Thanks @tyler-a-cox!

**Pull request merged**
